intg-denonavr/denon_remote.py

```diff
@@ -0,0 +1,820 @@
+# pylint: disable=C0302
+"""
+Remote entity functions.
+
+:license: Mozilla Public License Version 2.0, see LICENSE for more details.
+"""
+
+from typing import Any
+
+import avr
+import helpers
+import ucapi.remote
+from command_constants import (
+    AudysseyCommands,
+    CoreCommands,
+    DiracCommands,
+    SoundModeCommands,
+    VolumeCommands,
+)
+from config import AvrDevice, create_entity_id
+from media_player import DenonMediaPlayer
+from ucapi import EntityTypes, Remote, StatusCodes, media_player
+from ucapi.remote import Attributes, Commands, Features
+from ucapi.ui import Buttons
+
+# Mapping of an AVR state to a remote entity state
+REMOTE_STATE_MAPPING = {
+    avr.States.ON: ucapi.remote.States.ON,
+    avr.States.OFF: ucapi.remote.States.OFF,
+    avr.States.PAUSED: ucapi.remote.States.ON,
+    avr.States.PLAYING: ucapi.remote.States.ON,
+    avr.States.UNAVAILABLE: ucapi.remote.States.UNAVAILABLE,
+    avr.States.UNKNOWN: ucapi.remote.States.UNKNOWN,
+}
+
+
+# pylint: disable=R0903
+class DenonRemote(Remote):
+    """Representation of a Denon/Marantz AVR Remote entity."""
+
+    def __init__(self, device: AvrDevice, receiver: avr.DenonDevice, denon_media_player: DenonMediaPlayer):
+        """Initialize the class."""
+        self._device: avr.DenonDevice = receiver
+        self._denon_media_player: DenonMediaPlayer = denon_media_player
+        entity_id = create_entity_id(receiver.id, EntityTypes.REMOTE)
+        features = [Features.SEND_CMD, Features.ON_OFF, Features.TOGGLE]
+        super().__init__(
+            entity_id,
+            f"{device.name} Remote",
+            features,
+            attributes={
+                Attributes.STATE: receiver.state,
+            },
+            simple_commands=self._denon_media_player.get_supported_commands(),
+            button_mapping=REMOTE_BUTTONS_MAPPING,
+            ui_pages=DenonRemote._get_remote_ui_pages(),
+        )
+
+    async def command(self, cmd_id: str, params: dict[str, Any] | None = None) -> StatusCodes:
+        """
+        Remote entity command handler.
+
+        Called by the integration-API if a command is sent to a configured remote entity.
+
+        :param cmd_id: command
+        :param params: optional command parameters
+        :return: status code of the command request
+        """
+        if params:
+            repeat = self._get_int_param("repeat", params, 1)
+        else:
+            repeat = 1
+
+        is_succsess = False
+        for _ in range(0, repeat):
+            is_succsess |= await self._handle_command(cmd_id, params) == StatusCodes.OK
```

Comment:
Not sure what the best approach is for repeats: best-effort or fail-fast.
Usually I prefer fail-fast, but the chosen best-effort could work better for the intended use-case like "10 x volume down": a result of 9 x lowering is better than failing after the 1st failed command.
Therefore, let's leave it unless a better counter argument is found :-)

Reply:
I was thinking the same thing. I don't think you'd cause issues if some command fails, but it would be frustrating if everything would fail because one command didn't execute. The only downside is that you won't be made aware of potential issues. Maybe I should add logging?

Reply:
Okay, then I'll just keep the http logging